controller/object_store_controller.go

```diff
@@ -0,0 +1,870 @@
+// state, otherwise do nothing.
+func (osc *ObjectStoreController) handleRunning(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateStopped {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateStopping)
+		return nil
+	}
+
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil || dpl.Status.UnavailableReplicas > 0 {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err != nil {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	_, err = osc.ds.GetIngress(osc.namespace, store.Name)
+	if err != nil {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	pvc, err := osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err != nil || pvc.Status.Phase != corev1.ClaimBound {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	_, err = osc.ds.GetVolume(genPVName(store))
+	if err != nil {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolume(genPVName(store))
+	if err != nil {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopping(store *longhorn.ObjectStore) (err error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+	} else if *dpl.Spec.Replicas != 0 {
+		dpl.Spec.Replicas = int32Ptr(0)
+		_, err = osc.ds.UpdateDeployment(dpl)
+		return err
+	} else if dpl.Status.AvailableReplicas > 0 {
+		return nil // wait for shutdown
+	}
+
+	osc.setObjectStoreState(store, longhorn.ObjectStoreStateStopped)
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopped(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateRunning {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateStarting)
+		return nil
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore) (err error) {
+	// Just remove the finalizer. The K8s API will take care of removing dependent
+	// resources.
+	return osc.ds.RemoveFinalizerForObjectStore(store)
+}
+
+func (osc *ObjectStoreController) setObjectStoreState(
+	store *longhorn.ObjectStore,
+	status longhorn.ObjectStoreState,
+) (*longhorn.ObjectStore, error) {
+	store.Status.State = status
+	return osc.ds.UpdateObjectStoreStatus(store)
+}
+
+func (osc *ObjectStoreController) initializeObjectStore(store *longhorn.ObjectStore) (err error) {
+	// TODO: check if we really need/want to update the spec's image here
+	if store.Spec.Image == "" {
+		store.Spec.Image = osc.s3gwImage
+	}
+
+	if store.Spec.UiImage == "" {
+		store.Spec.UiImage = osc.uiImage
+	}
+
+	store.Spec.TargetState = longhorn.ObjectStoreStateRunning
```

Comment:
A common convention is that a controller should not modify the spec of its object but only the status. The default value should be set by webhook. See volume mutation webhook for example https://github.com/longhorn/longhorn-manager/blob/cf7b3e97346de9e677cf66081bf56cfcbb36d302/webhook/resources/volume/mutator.go#L159-L164

Reply:
I'll need to write a webhook whith the logic to update the `.spec.Image`, `.spec.UiImage` and `.spec.TargetState` properties then.